### PR TITLE
Make charging reducer typeclass

### PR DESCRIPTION
**Pull request by JosDenmark**
## Overview
This PR makes ChargingReducer into a typeclass in preparation of the F refactoring.

### JIRA ticket:
https://rchain.atlassian.net/browse/RHOL-1050

### Checklist
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by JosDenmark** [comment]
bors try

**Comment by JosDenmark** [comment]
@lukasz-golebiewski @odeac Reminder

**Comment by ArturGajowy** [comment]
I guess there are some compilation errors, and I see you've added more changes - please fix the compilation errors and ping us for another round of review.

I'm also not sure why there are two `make chargingReducer an implicit parameter` commits - probably should be just one, containing changes from both.

**Comment by JosDenmark** [comment]
They're for two different classes in two different files. One is for RholangOnlyDispatcher, the other is for RholangAndScalaDispatcher.

**Comment by JosDenmark** [comment]
bors try

**Comment by JosDenmark** [comment]
bors try

**Comment by JosDenmark** [comment]
@ArturGajowy @lukasz-golebiewski Ready for a re-review.

**Comment by JosDenmark** [comment]
bors try

**Comment by JosDenmark** [comment]
bors try

**Comment by JosDenmark** [comment]
bors try

**Comment by JosDenmark** [comment]
bors try

**Comment by JosDenmark** [comment]
bors try

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala [hunk 1]
```diff
@@ -174,7 +174,7 @@ class RuntimeManager[F[_]: Concurrent: ToAbstractContext] private (
             for {
               _              <- ToAbstractContext[F].fromTask(runtime.space.reset(hash))
               availablePhlos = Cost(deploy.raw.map(_.phloLimit).get)
-              _              <- ToAbstractContext[F].fromTask(runtime.reducer.setAvailablePhlos(availablePhlos))
+              _              <- ToAbstractContext[F].fromTask(runtime.reducer.setPhlo(availablePhlos))
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala]: again, for renames, the call-site changes should come together with the definition-site changes. In this case, the trait extracted in the first commit contains the renames. So either this goes into the first commit, or the first commit is cleaned-up to only contain the trait exxtraction (without method name/signature changes)

**Comment by ArturGajowy** [comment]
the `Merge branch 'dev' of rchain/rchain` commit is not needed (there were no conflicts, so the merge commit is 'empty'). Would a merge be necessary - github would let us know.

**Pull request merged**